### ISSUE TITLE
twisted.web.proxy.ReverseProxyResource incorrectly sends host header with no port

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#10728]
|<img alt="kragen's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| kragen reported|
|-|-|
|Trac ID|trac#1117|
|Type|defect|
|Created|2005-07-19 02:31:32Z|

Attachments:

* [nproxy.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/e6b/e6ba73dc9533e190c71bfc3877414870457f1533/0c8bd993a32aafb5c9500178d65b24cddf0b6c2a.py) (3363 bytes) - added by kragen on 2005-07-19 02:31:31Z - 
* [1117.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/e6b/e6ba73dc9533e190c71bfc3877414870457f1533/0775e4ecfbce37c988caa2f5dad2a84b175d3901.diff) (1055 bytes) - added by therve on 2007-12-29 14:58:43Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__1117 1117
type__defect defect
reporter__kragen kragen
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1121740292000000 1121740292000000
changetime__1199443173000000 1199443173000000
version__ 
owner__ 
cc__jknight cc__kragen cc__therve
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="kragen's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|kragen commented|
|-|-|

```
#!html
<pre>
This bug results in web servers proxied by
twisted.web.proxy.ReverseProxyResource generating incorrect redirect URLs if
they use the Host: HTTP header to construct their redirect URLs.  Apache has an
option to behave this way (although it's turned off by default) and my
installation of Jetty behaves this way.

This is related to issue 1109 in that both bugs relate to redirects generated by
origin servers, but this is a separate bug, and perhaps it is more clearly a bug.

According to http://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html#sec14.23 :
        A "host" without any trailing port information implies the
        default port for the service requested (e.g., "80" for an
        HTTP URL).

However, according to Twisted 2.0.1's
twisted.web.proxy.ReverseProxyResource.render, line 174:
        request.received_headers['host'] = self.host

This is mutating the request headers so it can pass them on to the backend
server.  However, it neglects to include self.port.  It would be correct to set
the host header to ('%s:%d' % (self.host, self.port)), but I think the following
is less surprising:

def hostport(host, port, defaultport=80):
    if port == defaultport: return host
    return '%s:%d' % (host, port)
...
        request.received_headers['host'] = hostport(self.host, self.port)

This leaves the behavior unchanged in the case where it was correct (at present,
twisted.web.proxy doesn't support https) and corrects it where it was wrong.

Attached is a file that includes this fix (in the form of a subclass of
ReverseProxyResource) and an end-to-end test that demonstrates its
effectiveness.  I apologize for not writing this test with twisted.trial; I
haven't yet learned twisted.trial, and I thought it would be better to submit
the fix first and rewrite the test afterwards.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="kragen's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|kragen commented|
|-|-|

```
#!html
<pre>
Perhaps I should have said, "this is related to #1109".
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to @therve|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve removed owner|
|-|-|

Now that there are tests, this is simpler :). Ready to review.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set owner to @therve|
|-|-|

Patch looks great.  Two suggestions. ;)

  * ```ReverseProxyResource.render``` has no docstring.  It could use one.
  * A reference to the section of the RFC kragen mentioned above, either in the implementation or near the new assertion in the test, would be really awesome.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve set status to `closed`|
|-|-|

(In [22258]) Fix twisted.web.proxy.ReverseProxyResource to specify the port in the 
host header if different from 80.

Authors: kragen, therve
Reviewer: exarkun
Fixes [#10728](https://github.com/twisted/twisted/issues/10728)

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
